File: Diffusion.m
function [ E, S ] = Diffusion( data, r, d, flag )
% Calculate the time-1 diffusion map for a pre-normalized data matrix.
%   INPUT
%       data : Data matrix (n x p).
%       r    : Number of neighbours (e.g. 15).
%       d    : Dimension of embedding (e.g. 8).
%   OUTPUT
%       E    : Diffusion embedding (n x d).
%       S    : Singular values (d + 1 x d + 1).
% Written by John Malik on 2018.6.23, john.malik@duke.edu.

if nargin < 4
    flag = false;
end

% number of points
n = size(data, 1);

% neighbour search
jj = knnsearch(data, data, 'k', r);
ii = transpose(1:n) * ones(1, r);

% adjacency matrix
A = sparse(ii, jj, ones(size(ii)), n, n);

% affinity matrix
if ~flag
    % number of neighbours x and y have in common
    W = A * A';
else W = A' * A; % number of pairs with x and y as common neighbours
end

% alpha normalize
D = sum(W, 2);
W = bsxfun(@rdivide, bsxfun(@rdivide, W, D), transpose(D));

% normalized graph laplacian
D = sqrt(sum(W, 2));
P = bsxfun(@rdivide, bsxfun(@rdivide, W, D), transpose(D));

% eigendecomposition
opts.issym = 1;
opts.isreal = 1;
opts.v0 = ones(n, 1);
[E, S] = eigs((P + P') / 2, d + 1, 'la', opts);

% shift back and normalize
E = bsxfun(@rdivide, E(:, 2:end), D);
E = E * S(2:end, 2:end);

end

File: Diffusion.py
import numpy as np
from sklearn.neighbors import NearestNeighbors
from scipy import sparse

# data is an n x p matrix with rows as observations
# k is the number of neighbours (e.g. 20)
# d is the dimension of the embedding (e.g. 3)
# this implementation uses the zero-one kernel and diffusion time 0

def Diffusion( data, k, d ):

    # number of points
    n = data.shape[0];

    # neighbour search (replace with your favourite algorithm)
    knn = NearestNeighbors(n_neighbors=k).fit(data)
    _, jj = knn.kneighbors(data)
    ii = np.repeat(range(0, n), k)

    # affinity matrix
    W = sparse.coo_matrix((np.ones(n * k), (ii, jj.ravel())))
    W = W * W.transpose()

    # alpha normalization
    D = sparse.diags(1 / W.sum(axis=1).A.ravel())
    W = D * W * D

    # normalized graph laplacian
    D = sparse.diags(1 / np.sqrt(W.sum(axis=1)).A.ravel())
    P = D * W * D

    # eigendecomposition
    _, E = sparse.linalg.eigsh((P + P.transpose()) / 2, k=d+1, which='LM')

    # shift back and normalize
    E = D * E[:, :-1]
    # E = np.divide(E, np.linalg.norm(E, axis=0))

    return E

# E is an n x d matrix with rows as observations
# plot the last three columns of E for visualization purposes

# Written by John Malik on 2018.6.23, john.malik@duke.edu.
